File: bench/diff_bitmap-str.c
// SPDX-License-Identifier: GPL-2.0
// Tier-2.5 differential oracle: C original vs Rust translation, bitmap-str.
// Reference extracted from lib/bitmap-str.c (v7.1); kept byte-identical
// for the translated subset (bitmap_parselist, bitmap_parse, and their
// static string-parsing helpers — the *_user/print_* family stays C-
// only, not exercised here).
#include <ctype.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef unsigned long ul;
#define BITS_PER_LONG (sizeof(ul) * 8)
#define BITS_TO_LONGS(nr) (((nr) + BITS_PER_LONG - 1) / BITS_PER_LONG)
#define BITS_TO_U32(nr) (((nr) + 31) / 32)
#define EINVAL 22
#define ERANGE 34
#define EOVERFLOW 75
#define MAX_ERRNO 4095

static void *err_ptr(long e) { return (void *)e; }
static long ptr_err(const void *p) { return (long)p; }
static int is_err(const void *p) { return (unsigned long)p >= (unsigned long)-MAX_ERRNO; }

static int my_hex_to_bin(unsigned char ch)
{
	unsigned char cu = ch & 0xdf;
	return -1 +
	       ((ch - '0' + 1) & (unsigned)((ch - '9' - 1) & ('0' - 1 - ch)) >> 8) +
	       ((cu - 'A' + 11) & (unsigned)((cu - 'F' - 1) & ('A' - 1 - cu)) >> 8);
}

// _parse_integer (lib/kstrtox.c, already validated in bench/diff_kstrtox.c)
static unsigned int my_parse_integer(const char *s, unsigned int base, uint64_t *p)
{
	uint64_t res = 0;
	unsigned int rv = 0;
	int max_chars = 2147483647;
	while (max_chars--) {
		unsigned int c = (unsigned char)*s;
		unsigned int lc = c | 0x20;
		unsigned int val;
		if ('0' <= c && c <= '9') val = c - '0';
		else if ('a' <= lc && lc <= 'f') val = lc - 'a' + 10;
		else break;
		if (val >= base) break;
		if (res & (~0ull << 60)) {
			if (res > (UINT64_MAX - val) / base) rv |= (1U << 31);
		}
		res = res * base + val;
		rv++;
		s++;
	}
	*p = res;
	return rv;
}

static void my_bitmap_zero(ul *dst, unsigned int nbits)
{ memset(dst, 0, BITS_TO_LONGS(nbits) * sizeof(ul)); }

static void my_bitmap_set(ul *map, unsigned int start, int len)
{
	ul *p = map + start / BITS_PER_LONG;
	const unsigned int size = start + len;
	int bits_to_set = BITS_PER_LONG - (start % BITS_PER_LONG);
	ul mask_to_set = ~0UL << (start & (BITS_PER_LONG - 1));
	while (len - bits_to_set >= 0) {
		*p |= mask_to_set;
		len -= bits_to_set;
		bits_to_set = BITS_PER_LONG;
		mask_to_set = ~0UL;
		p++;
	}
	if (len) {
		mask_to_set &= (~0UL >> (-size & (BITS_PER_LONG - 1)));
		*p |= mask_to_set;
	}
}

static void my_bitmap_clear(ul *map, unsigned int start, int len)
{
	ul *p = map + start / BITS_PER_LONG;
	const unsigned int size = start + len;
	int bits_to_clear = BITS_PER_LONG - (start % BITS_PER_LONG);
	ul mask_to_clear = ~0UL << (start & (BITS_PER_LONG - 1));
	while (len - bits_to_clear >= 0) {
		*p &= ~mask_to_clear;
		len -= bits_to_clear;
		bits_to_clear = BITS_PER_LONG;
		mask_to_clear = ~0UL;
		p++;
	}
	if (len) {
		mask_to_clear &= (~0UL >> (-size & (BITS_PER_LONG - 1)));
		*p &= ~mask_to_clear;
	}
}

static ul my_find_next_bit(const ul *addr, ul size, ul offset)
{
	for (ul i = offset; i < size; i++)
		if (addr[i / BITS_PER_LONG] & (1UL << (i % BITS_PER_LONG))) return i;
	return size;
}

static int my_end_of_str(char c) { return c == '\0' || c == '\n'; }
static int my_end_of_region_(char c) { return isspace((unsigned char)c) || c == ','; }
static int my_end_of_region(char c) { return my_end_of_region_(c) || my_end_of_str(c); }

static const char *my_bitmap_getnum(const char *str, unsigned int *num, unsigned int lastbit)
{
	uint64_t n;
	unsigned int len;
	if (str[0] == 'N') { *num = lastbit; return str + 1; }
	len = my_parse_integer(str, 10, &n);
	if (!len) return err_ptr(-EINVAL);
	if (len & (1U << 31) || n != (unsigned int)n) return err_ptr(-EOVERFLOW);
	*num = n;
	return str + len;
}

static const char *my_bitmap_find_region(const char *str)
{
	while (my_end_of_region_(*str)) str++;
	return my_end_of_str(*str) ? NULL : str;
}

static const char *my_bitmap_find_region_reverse(const char *start, const char *end)
{
	while (start <= end && my_end_of_region_(*end)) end--;
	return end;
}

struct region { unsigned int start, off, group_len, end, nbits; };

static const char *my_bitmap_parse_region(const char *str, struct region *r)
{
	unsigned int lastbit = r->nbits - 1;
	if (!strncasecmp(str, "all", 3)) {
		r->start = 0; r->end = lastbit; str += 3;
		goto check_pattern;
	}
	str = my_bitmap_getnum(str, &r->start, lastbit);
	if (is_err(str)) return str;
	if (my_end_of_region(*str)) goto no_end;
	if (*str != '-') return err_ptr(-EINVAL);
	str = my_bitmap_getnum(str + 1, &r->end, lastbit);
	if (is_err(str)) return str;
check_pattern:
	if (my_end_of_region(*str)) goto no_pattern;
	if (*str != ':') return err_ptr(-EINVAL);
	str = my_bitmap_getnum(str + 1, &r->off, lastbit);
	if (is_err(str)) return str;
	if (*str != '/') return err_ptr(-EINVAL);
	return my_bitmap_getnum(str + 1, &r->group_len, lastbit);
no_end:
	r->end = r->start;
no_pattern:
	r->off = r->end + 1;
	r->group_len = r->end + 1;
	return my_end_of_str(*str) ? NULL : str;
}

static void my_bitmap_set_region(const struct region *r, ul *bitmap)
{
	unsigned int start;
	for (start = r->start; start <= r->end; start += r->group_len) {
		unsigned int l = r->end - start + 1;
		if (r->off < l) l = r->off;
		my_bitmap_set(bitmap, start, l);
	}
}

static int my_bitmap_check_region(const struct region *r)
{
	if (r->start > r->end || r->group_len == 0 || r->off > r->group_len) return -EINVAL;
	if (r->end >= r->nbits) return -ERANGE;
	return 0;
}

static int my_bitmap_parselist(const char *buf, ul *maskp, int nmaskbits)
{
	struct region r;
	long ret;
	r.nbits = nmaskbits;
	my_bitmap_zero(maskp, r.nbits);
	while (buf) {
		buf = my_bitmap_find_region(buf);
		if (buf == NULL) return 0;
		buf = my_bitmap_parse_region(buf, &r);
		if (is_err(buf)) return ptr_err(buf);
		ret = my_bitmap_check_region(&r);
		if (ret) return ret;
		my_bitmap_set_region(&r, maskp);
	}
	return 0;
}

static const char *my_bitmap_get_x32_reverse(const char *start, const char *end, uint32_t *num)
{
	uint32_t ret = 0;
	int c, i;
	for (i = 0; i < 32; i += 4) {
		c = my_hex_to_bin((unsigned char)*end--);
		if (c < 0) return err_ptr(-EINVAL);
		ret |= (uint32_t)c << i;
		if (start > end || my_end_of_region_(*end)) goto out;
	}
	if (my_hex_to_bin((unsigned char)*end--) >= 0) return err_ptr(-EOVERFLOW);
out:
	*num = ret;
	return end;
}

static int my_bitmap_parse(const char *start, unsigned int buflen, ul *maskp, int nmaskbits)
{
	const char *nl = memchr(start, '\n', buflen);
	const char *end = (nl ? nl : start + strnlen(start, buflen)) - 1;
	int chunks = BITS_TO_U32(nmaskbits);
	uint32_t *bitmap = (uint32_t *)maskp;
	int chunk;
	for (chunk = 0; ; chunk++) {
		end = my_bitmap_find_region_reverse(start, end);
		if (start > end) break;
		if (!chunks--) return -EOVERFLOW;
		end = my_bitmap_get_x32_reverse(start, end, &bitmap[chunk]);
		if (is_err(end)) return ptr_err(end);
	}
	int unset_bit = (BITS_TO_U32(nmaskbits) - chunks) * 32;
	if (unset_bit < nmaskbits) {
		my_bitmap_clear(maskp, unset_bit, nmaskbits - unset_bit);
		return 0;
	}
	if (my_find_next_bit(maskp, unset_bit, nmaskbits) != (ul)unset_bit) return -EOVERFLOW;
	return 0;
}

// Explicit LCG (same constants as bench/diff_base64.c).
static uint64_t lcg_state;
static uint32_t lcg_next(void)
{
	lcg_state = lcg_state * 6364136223846793005ULL + 1442695040888963407ULL;
	return (uint32_t)(lcg_state >> 32);
}

#define NBITS 256
#define NWORDS (NBITS / 64)

// Generate a plausible bitmap-list string: comma-separated
// numbers/ranges/patterns, occasionally "all" or "N".
static void gen_listbuf(char *buf, int maxlen)
{
	int pos = 0;
	int nterms = 1 + lcg_next() % 5;
	for (int t = 0; t < nterms && pos < maxlen - 20; t++) {
		if (t > 0) buf[pos++] = ',';
		int kind = lcg_next() % 6;
		if (kind == 0) {
			pos += sprintf(buf + pos, "all");
		} else {
			int a = lcg_next() % NBITS;
			if (lcg_next() % 3 == 0 && kind != 5) {
				pos += sprintf(buf + pos, "N");
			} else {
				pos += sprintf(buf + pos, "%d", a);
			}
			if (kind >= 2) {
				int b = lcg_next() % NBITS;
				pos += sprintf(buf + pos, "-%d", b);
				if (kind >= 4) {
					int off = 1 + lcg_next() % 4;
					int grp = off + lcg_next() % 4;
					pos += sprintf(buf + pos, ":%d/%d", off, grp);
				}
			}
		}
	}
	buf[pos] = 0;
}

static void gen_hexbuf(char *buf, int maxlen)
{
	int len = 1 + lcg_next() % (maxlen - 1);
	static const char hexdigits[] = "0123456789abcdefABCDEF";
	int pos = 0;
	for (int i = 0; i < len && pos < maxlen - 1; i++) {
		if (i > 0 && lcg_next() % 8 == 0) buf[pos++] = ',';
		else buf[pos++] = hexdigits[lcg_next() % (sizeof(hexdigits) - 1)];
	}
	buf[pos] = 0;
}

int main(int argc, char **argv)
{
	long n = argc > 1 ? atol(argv[1]) : 5000;
	lcg_state = argc > 2 ? (uint64_t)atol(argv[2]) : 271828;

	for (long i = 0; i < n; i++) {
		char listbuf[128];
		gen_listbuf(listbuf, sizeof(listbuf) - 1);
		ul mask[NWORDS];
		int r1 = my_bitmap_parselist(listbuf, mask, NBITS);
		printf("parselist,%s,%d,%016lx,%016lx\n", listbuf, r1, mask[0], mask[NWORDS - 1]);

		char hexbuf[80];
		gen_hexbuf(hexbuf, sizeof(hexbuf) - 1);
		ul mask2[NWORDS] = {0};
		int r2 = my_bitmap_parse(hexbuf, (unsigned int)strlen(hexbuf), mask2, NBITS);
		printf("parse,%s,%d,%016lx,%016lx\n", hexbuf, r2, mask2[0], mask2[NWORDS - 1]);
	}
	return 0;
}

File: bench/diff_bitmap-str.rs
// SPDX-License-Identifier: GPL-2.0
//! Tier-2.5 differential oracle: Rust translation side, bitmap-str.
//! Faithful copy of lib/bitmap-str_rs.rs's algorithms (kernel-crate
//! bindings/export stripped, cross-TU calls to kstrtox/bitmap/find_bit/
//! string/hexdump inlined as local fns) — same protocol/LCG as
//! diff_bitmap-str.c. Uses real raw pointers (not offsets) so the
//! ERR_PTR/NULL sentinel encoding matches the actual translation
//! exactly — a prior offset-based version spuriously flagged
//! legitimate "one before buffer start" positions as ERR_PTRs.

const BITS_PER_LONG: u32 = usize::BITS;
const MAX_ERRNO: isize = 4095;
const EINVAL: i32 = 22;
const ERANGE: i32 = 34;
const EOVERFLOW: i32 = 75;
const KSTRTOX_OVERFLOW: u32 = 1u32 << 31;
const NBITS: usize = 256;
const NWORDS: usize = NBITS / 64;

unsafe fn err_ptr(error: isize) -> *const u8 {
    error as *const u8
}
unsafe fn ptr_err(ptr: *const u8) -> isize {
    ptr as isize
}
unsafe fn is_err(ptr: *const u8) -> bool {
    (ptr as usize) >= (-MAX_ERRNO) as usize
}

// _parse_integer (lib/kstrtox_rs.rs), base fixed to 10 here (only
// caller in this file is bitmap_getnum).
unsafe fn parse_integer(mut s: *const u8) -> (usize, u64) {
    let mut res: u64 = 0;
    let mut rv: usize = 0;
    loop {
        let c = unsafe { *s };
        if !c.is_ascii_digit() {
            break;
        }
        let val = (c - b'0') as u64;
        if res & (!0u64 << 60) != 0 && res > (u64::MAX - val) / 10 {
            rv |= KSTRTOX_OVERFLOW as usize;
        }
        res = res.wrapping_mul(10).wrapping_add(val);
        rv += 1;
        s = unsafe { s.add(1) };
    }
    (rv, res)
}

fn hex_to_bin(ch: u8) -> i32 {
    let cu = ch & 0xdf;
    -1 + ((((ch.wrapping_sub(b'0').wrapping_add(1)) as u32)
        & (((ch.wrapping_sub(b'9').wrapping_sub(1)) as i8 as i32
            & (b'0' as i32 - 1 - ch as i32)) as u32
            >> 8)) as i32)
        + ((((cu.wrapping_sub(b'A').wrapping_add(11)) as u32)
            & (((cu.wrapping_sub(b'F').wrapping_sub(1)) as i8 as i32
                & (b'A' as i32 - 1 - cu as i32)) as u32
                >> 8)) as i32)
}

fn bitmap_zero(dst: &mut [u64]) {
    for w in dst.iter_mut() {
        *w = 0;
    }
}

fn bitmap_set(map: &mut [u64], start: u32, len: i32) {
    let mut p = (start / BITS_PER_LONG) as usize;
    let size = start.wrapping_add(len as u32);
    let mut bits_to_set = BITS_PER_LONG - (start % BITS_PER_LONG);
    let mut mask_to_set: u64 = !0u64 << (start & (BITS_PER_LONG - 1));
    let mut len = len;
    while len - bits_to_set as i32 >= 0 {
        map[p] |= mask_to_set;
        len -= bits_to_set as i32;
        bits_to_set = BITS_PER_LONG;
        mask_to_set = !0u64;
        p += 1;
    }
    if len != 0 {
        mask_to_set &= !0u64 >> ((0u32.wrapping_sub(size)) & (BITS_PER_LONG - 1));
        map[p] |= mask_to_set;
    }
}

fn bitmap_clear(map: &mut [u64], start: u32, len: i32) {
    let mut p = (start / BITS_PER_LONG) as usize;
    let size = start.wrapping_add(len as u32);
    let mut bits_to_clear = BITS_PER_LONG - (start % BITS_PER_LONG);
    let mut mask_to_clear: u64 = !0u64 << (start & (BITS_PER_LONG - 1));
    let mut len = len;
    while len - bits_to_clear as i32 >= 0 {
        map[p] &= !mask_to_clear;
        len -= bits_to_clear as i32;
        bits_to_clear = BITS_PER_LONG;
        mask_to_clear = !0u64;
        p += 1;
    }
    if len != 0 {
        mask_to_clear &= !0u64 >> ((0u32.wrapping_sub(size)) & (BITS_PER_LONG - 1));
        map[p] &= !mask_to_clear;
    }
}

fn find_next_bit(addr: &[u64], size: u32, offset: u32) -> u32 {
    let mut i = offset;
    while i < size {
        if addr[(i / BITS_PER_LONG) as usize] & (1u64 << (i % BITS_PER_LONG)) != 0 {
            return i;
        }
        i += 1;
    }
    size
}

fn end_of_str(c: u8) -> bool {
    c == 0 || c == b'\n'
}
fn end_of_region_(c: u8) -> bool {
    c.is_ascii_whitespace() || c == b','
}
fn end_of_region(c: u8) -> bool {
    end_of_region_(c) || end_of_str(c)
}

unsafe fn bitmap_getnum(str_: *const u8, num: *mut u32, lastbit: u32) -> *const u8 {
    unsafe {
        if *str_ == b'N' {
            *num = lastbit;
            return str_.add(1);
        }
        let (len, n) = parse_integer(str_);
        if len == 0 {
            return err_ptr(-(EINVAL as isize));
        }
        if (len as u32) & KSTRTOX_OVERFLOW != 0 || n != n as u32 as u64 {
            return err_ptr(-(EOVERFLOW as isize));
        }
        *num = n as u32;
        str_.add(len)
    }
}

unsafe fn bitmap_find_region(str_: *const u8) -> *const u8 {
    let mut str_ = str_;
    unsafe {
        while end_of_region_(*str_) {
            str_ = str_.add(1);
        }
        if end_of_str(*str_) {
            core::ptr::null()
        } else {
            str_
        }
    }
}

unsafe fn bitmap_find_region_reverse(start: *const u8, end: *const u8) -> *const u8 {
    let mut end = end;
    unsafe {
        while start <= end && end_of_region_(*end) {
            end = end.sub(1);
        }
        end
    }
}

struct Region {
    start: u32,
    off: u32,
    group_len: u32,
    end: u32,
    nbits: u32,
}

unsafe fn strncasecmp_all(s: *const u8) -> bool {
    unsafe {
        (*s | 0x20) == b'a'
            && (*s.add(1) | 0x20) == b'l'
            && (*s.add(2) | 0x20) == b'l'
    }
}

unsafe fn bitmap_parse_region(str_: *const u8, r: *mut Region) -> *const u8 {
    unsafe {
        let lastbit = (*r).nbits - 1;
        let mut str_ = str_;

        if strncasecmp_all(str_) {
            (*r).start = 0;
            (*r).end = lastbit;
            str_ = str_.add(3);
        } else {
            str_ = bitmap_getnum(str_, &mut (*r).start, lastbit);
            if is_err(str_) {
                return str_;
            }

            if end_of_region(*str_) {
                (*r).end = (*r).start;
                (*r).off = (*r).end + 1;
                (*r).group_len = (*r).end + 1;
                return if end_of_str(*str_) { core::ptr::null() } else { str_ };
            }

            if *str_ != b'-' {
                return err_ptr(-(EINVAL as isize));
            }
            str_ = bitmap_getnum(str_.add(1), &mut (*r).end, lastbit);
            if is_err(str_) {
                return str_;
            }
        }

        // check_pattern:
        if end_of_region(*str_) {
            (*r).off = (*r).end + 1;
            (*r).group_len = (*r).end + 1;
            return if end_of_str(*str_) { core::ptr::null() } else { str_ };
        }
        if *str_ != b':' {
            return err_ptr(-(EINVAL as isize));
        }
        str_ = bitmap_getnum(str_.add(1), &mut (*r).off, lastbit);
        if is_err(str_) {
            return str_;
        }
        if *str_ != b'/' {
            return err_ptr(-(EINVAL as isize));
        }
        bitmap_getnum(str_.add(1), &mut (*r).group_len, lastbit)
    }
}

fn bitmap_set_region(r: &Region, bitmap: &mut [u64]) {
    let mut start = r.start;
    while start <= r.end {
        let len = core::cmp::min(r.end - start + 1, r.off);
        bitmap_set(bitmap, start, len as i32);
        start += r.group_len;
    }
}

fn bitmap_check_region(r: &Region) -> i32 {
    if r.start > r.end || r.group_len == 0 || r.off > r.group_len {
        return -EINVAL;
    }
    if r.end >= r.nbits {
        return -ERANGE;
    }
    0
}

unsafe fn bitmap_parselist(buf: *const u8, maskp: &mut [u64], nmaskbits: i32) -> i32 {
    let mut r = Region { start: 0, off: 0, group_len: 0, end: 0, nbits: nmaskbits as u32 };
    bitmap_zero(maskp);

    unsafe {
        let mut buf = buf;
        loop {
            if buf.is_null() {
                return 0;
            }
            buf = bitmap_find_region(buf);
            if buf.is_null() {
                return 0;
            }

            buf = bitmap_parse_region(buf, &mut r);
            if is_err(buf) {
                return ptr_err(buf) as i32;
            }

            let ret = bitmap_check_region(&r);
            if ret != 0 {
                return ret;
            }

            bitmap_set_region(&r, maskp);
        }
    }
}

fn bits_to_u32(nr: u32) -> u32 {
    nr.div_ceil(32)
}

unsafe fn bitmap_get_x32_reverse(start: *const u8, end: *const u8, num: *mut u32) -> *const u8 {
    let mut ret: u32 = 0;
    let mut end = end;
    unsafe {
        let mut i = 0;
        while i < 32 {
            let c = hex_to_bin(*end);
            end = end.sub(1);
            if c < 0 {
                return err_ptr(-(EINVAL as isize));
            }
            ret |= (c as u32) << i;

            if start > end || end_of_region_(*end) {
                *num = ret;
                return end;
            }
            i += 4;
        }
        if hex_to_bin(*end) >= 0 {
            return err_ptr(-(EOVERFLOW as isize));
        }
        *num = ret;
        end
    }
}

unsafe fn bitmap_parse(start: *const u8, buflen: u32, maskp: &mut [u64], nmaskbits: i32) -> i32 {
    unsafe {
        // strnchrnul(start, buflen, '\n') - 1
        let mut p = start;
        let mut n = 0u32;
        while n < buflen && *p != 0 && *p != b'\n' {
            p = p.add(1);
            n += 1;
        }
        let mut end = p.sub(1);

        let mut chunks = bits_to_u32(nmaskbits as u32) as i32;
        let bitmap32: &mut [u32] = bytemuck_u64_to_u32_mut(maskp);
        let mut chunk: isize = 0;

        loop {
            end = bitmap_find_region_reverse(start, end);
            if start > end {
                break;
            }

            if chunks == 0 {
                return -EOVERFLOW;
            }
            chunks -= 1;

            let r = bitmap_get_x32_reverse(start, end, &mut bitmap32[chunk as usize]);
            if is_err(r) {
                return ptr_err(r) as i32;
            }
            end = r;
            chunk += 1;
        }

        let unset_bit = (bits_to_u32(nmaskbits as u32) as i32 - chunks) * 32;
        if unset_bit < nmaskbits {
            bitmap_clear(maskp, unset_bit as u32, nmaskbits - unset_bit);
            return 0;
        }
        if find_next_bit(maskp, nmaskbits as u32, unset_bit as u32) != unset_bit as u32 {
            return -EOVERFLOW;
        }
        0
    }
}

fn bytemuck_u64_to_u32_mut(s: &mut [u64]) -> &mut [u32] {
    // SAFETY: u64 slice is always validly reinterpretable as twice as
    // many u32s on this (little-endian) host, matching the C side's
    // `(uint32_t *)maskp` cast under LE.
    unsafe { core::slice::from_raw_parts_mut(s.as_mut_ptr().cast::<u32>(), s.len() * 2) }
}

struct Lcg(u64);
impl Lcg {
    fn next(&mut self) -> u32 {
        self.0 = self.0.wrapping_mul(6364136223846793005).wrapping_add(1442695040888963407);
        (self.0 >> 32) as u32
    }
}

fn gen_listbuf(rng: &mut Lcg, maxlen: usize) -> Vec<u8> {
    let mut buf = Vec::with_capacity(maxlen + 1);
    let nterms = 1 + rng.next() % 5;
    for t in 0..nterms {
        if buf.len() >= maxlen.saturating_sub(20) {
            break;
        }
        if t > 0 {
            buf.push(b',');
        }
        let kind = rng.next() % 6;
        if kind == 0 {
            buf.extend_from_slice(b"all");
        } else {
            let a = rng.next() % NBITS as u32;
            if rng.next() % 3 == 0 && kind != 5 {
                buf.push(b'N');
            } else {
                buf.extend_from_slice(a.to_string().as_bytes());
            }
            if kind >= 2 {
                let b = rng.next() % NBITS as u32;
                buf.push(b'-');
                buf.extend_from_slice(b.to_string().as_bytes());
                if kind >= 4 {
                    let off = 1 + rng.next() % 4;
                    let grp = off + rng.next() % 4;
                    buf.push(b':');
                    buf.extend_from_slice(off.to_string().as_bytes());
                    buf.push(b'/');
                    buf.extend_from_slice(grp.to_string().as_bytes());
                }
            }
        }
    }
    buf.push(0);
    buf
}

fn gen_hexbuf(rng: &mut Lcg, maxlen: usize) -> Vec<u8> {
    const HEXDIGITS: &[u8] = b"0123456789abcdefABCDEF";
    let len = 1 + (rng.next() as usize) % (maxlen - 1);
    let mut buf = Vec::with_capacity(maxlen + 1);
    for i in 0..len {
        if buf.len() >= maxlen - 1 {
            break;
        }
        if i > 0 && rng.next() % 8 == 0 {
            buf.push(b',');
        } else {
            buf.push(HEXDIGITS[(rng.next() as usize) % HEXDIGITS.len()]);
        }
    }
    buf.push(0);
    buf
}

fn cstr(v: &[u8]) -> &str {
    let end = v.iter().position(|&b| b == 0).unwrap_or(v.len());
    std::str::from_utf8(&v[..end]).unwrap()
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let n: u64 = args.get(1).and_then(|a| a.parse().ok()).unwrap_or(5000);
    let seed: u64 = args.get(2).and_then(|a| a.parse().ok()).unwrap_or(271828);
    let mut rng = Lcg(seed);

    for _ in 0..n {
        let listbuf = gen_listbuf(&mut rng, 127);
        let mut mask = [0u64; NWORDS];
        let r1 = unsafe { bitmap_parselist(listbuf.as_ptr(), &mut mask, NBITS as i32) };
        println!("parselist,{},{},{:016x},{:016x}", cstr(&listbuf), r1, mask[0], mask[NWORDS - 1]);

        let hexbuf = gen_hexbuf(&mut rng, 79);
        let mut mask2 = [0u64; NWORDS];
        let hexlen = hexbuf.iter().position(|&b| b == 0).unwrap_or(hexbuf.len());
        let r2 = unsafe { bitmap_parse(hexbuf.as_ptr(), hexlen as u32, &mut mask2, NBITS as i32) };
        println!("parse,{},{},{:016x},{:016x}", cstr(&hexbuf), r2, mask2[0], mask2[NWORDS - 1]);
    }
}
